[RFC] Device Array and compat Device Vector (#2660)
uxlfoundation/oneDPL: rfcs/proposed/device_vector/README.md

| Implementation | Source |
|---|---|
| **Thrust** (`thrust::device_vector`) | [NVIDIA/cccl - device_vector.h](https://github.com/NVIDIA/cccl/blob/main/thrust/thrust/device_vector.h) |
| **SYCLomatic** (`dpct::device_vector`) | [SYCLomatic - vector.h](https://github.com/oneapi-src/SYCLomatic/blob/SYCLomatic/clang/runtime/dpct-rt/include/dpct/dpl_extras/vector.h) |
| **Distributed Ranges** (`dr::sp::device_vector`) | [distributed-ranges - device_vector.hpp](https://github.com/oneapi-src/distributed-ranges/blob/main/include/dr/sp/device_vector.hpp) |
| **sycl-thrust** (`thrust::device_vector`) | [SparseBLAS/sycl-thrust - device_vector.h](https://github.com/SparseBLAS/sycl-thrust/blob/main/include/thrust/device_vector.h) |

### How They Differ

| Aspect | Proposed (oneDPL) | Thrust | sycl-thrust | SYCLomatic |
|---|---|---|---|---|
| **Default Allocator** | `device_allocator<T>` wrapping `sycl::malloc_device`; custom `DeviceAllocator` concept | `thrust::device_allocator<T>` (CUDA `cudaMalloc`) | `device_allocator<T>` (`sycl::malloc_device`); supports alignment template parameter | USM: `sycl::usm_allocator<T, shared>` / Buffer: `__buffer_allocator<T>` |

> | **Memory Model** | **Device memory** via `sycl::malloc_device`; host access triggers explicit transfers | **Device memory** via `cudaMalloc`; host access triggers explicit transfers | **Device memory** via `sycl::malloc_device`; explicit transfers | **Shared memory** via USM shared or SYCL buffer/accessor; runtime manages placement |

Reviewer: Is "host access triggers explicit transfers" accurate? I mean, if something is triggered by another action, it is implicit and not explicit, no?

Author: fixed

  queue and imply synchronization semantics. Queues are accepted per-operation
  or created on demand.

- **Type T should only require device copyability.**
  We should not need anything except device copyability (for copy to and from
  the device).

- **No tag system for dispatch to specific hardware.**
  Execution policies dictate where algorithms are run. We don't intend to
  provide other flavors of vector / iterator which would have different tags,
  which would be required to dispatch based upon tag.

- **Custom `DeviceAllocator` concept for pluggable allocation.**
  A minimal allocator interface — just `allocate(n, ctx, dev)` and
  `deallocate(p, n, ctx, dev)` — that avoids the `std::allocator` named
  requirements (which mandate host-accessible memory). Enables pool allocators,
  aligned allocation, and other strategies. See the
  [device_array allocator section](device_array.md#allocator) for details.

- **No `push_back`, `insert`, `erase`.**
  Rarely used in practice (see [usage study](usage_pattern_study.md)),
  high implementation complexity for device memory.

> - **Host-side operations block but do not synchronize with prior work.**

Reviewer: What does "block" mean in this sentence, and is it important in the context of design decisions?

Author: It means they `wait()` until the operation is complete before returning, but no guarantee that previously queue work finishes before this starts.  This is mentioned because it is a departure from the `thrust::device_vector`.  I can make this more clear, and provide better justification for why this decision is made (fitting better with sycl, and also from the usage study showing user's preferences)

  The user is responsible for ensuring prior kernels have completed before
  host-side access. This can be achieved via an in-order queue or explicit
  event waits. `device_array` additionally offers async overloads with
  `depends_on` events.

## Open Questions

- **Should `device_array`'s async overloads be in the initial release or
  deferred?**
  see [device_array](device_array.md).

- **Header organization?**
  - We could have a `<oneapi/dpl/compat>` header and automatically include `device_array` with other includes?
  Alternatives:
  - Individual headers:
  `<oneapi/dpl/device_array>` and `<oneapi/dpl/device_vector>`.. `device_vector`
  would transitively include `device_array` since it depends on it.

>   - We could have a `compat` header and a individual `device_array` header. However, if we intend to use `device_array` within our own sycl implementations, that may impact our decision here.

Reviewer: This is not quite clear to me. What do you mean by "within our own sycl implementation", why would we want to use `device_array` there, and how that impacts the decision about headers?

Author: I've reworded this to be more clear. I mean that if oneDPL wants to use `device_array` as temporary storage within oneDPL's sycl backend implementation, that may impact where we want the header to live.

Just like users, we may gain some benefits from this easy allocation / lifetime management and use it as an alternative to `sycl::buffer`.

# `device_array` and `compat::device_vector` for oneDPL

## Introduction

This RFC proposes adding a container to oneDPL that provides

> a `std::vector`-like interface for managing device memory.

Reviewer: ```suggestion
This RFC proposes adding data containers to oneDPL for managing device memory and data transfer.
```

### Motivation

- **Migration from CUDA/Thrust** - Thrust's `device_vector` is heavily used
  in CUDA codebases. Providing an equivalent in oneDPL lowers the barrier
  for porting to SYCL backends. SYCLomatic already generates code targeting
  a `dpct::device_vector` compatibility shim, and having an official oneDPL
  type would give that migration a stable target, in a repository which is
  actively maintained.
- **Ease of use** - Users currently must manually manage USM allocations or
  SYCL buffers and pair them with raw pointers or iterators. A

>   `device_vector` encapsulates allocation, sizing, and lifetime in a

Reviewer: ```suggestion
  device container encapsulates allocation, sizing, and lifetime in a
```

     operations across all domains. `device_vector` is primarily used as an
     RAII device memory manager and host-device data shuttle.
  2. **`begin()`/`end()` integration with parallel algorithms** is the
     second-most critical capability.
  3. **Some popular AI/ML projects** (FAISS, cuDF, cuML) have **moved away from
     `thrust::device_vector`** due to unwanted value initialization, lack of
     stream parameters, and header bloat — then built alternatives that
     prioritize explicit async control and uninitialized allocation. Other HPC
     and ML projects remain heavy users.
  4. **Full `std::vector`-like modifiers** (`push_back`, `insert`, `erase`)
     are rarely used in real workloads.

## Comparison of Existing device_vector Implementations

| Implementation | Source |
|---|---|
| **Thrust** (`thrust::device_vector`) | [NVIDIA/cccl - device_vector.h](https://github.com/NVIDIA/cccl/blob/main/thrust/thrust/device_vector.h) |
| **SYCLomatic** (`dpct::device_vector`) | [SYCLomatic - vector.h](https://github.com/oneapi-src/SYCLomatic/blob/SYCLomatic/clang/runtime/dpct-rt/include/dpct/dpl_extras/vector.h) |

> | **Distributed Ranges** (`dr::sp::device_vector`) | [distributed-ranges - device_vector.hpp](https://github.com/oneapi-src/distributed-ranges/blob/main/include/dr/sp/device_vector.hpp) |

Reviewer: It is not used in the analysis below. Remove?

Author: removed

| **sycl-thrust** (`thrust::device_vector`) | [SparseBLAS/sycl-thrust - device_vector.h](https://github.com/SparseBLAS/sycl-thrust/blob/main/include/thrust/device_vector.h) |

### How They Differ

| Aspect | Proposed (oneDPL) | Thrust | sycl-thrust | SYCLomatic |
|---|---|---|---|---|
| **Default Allocator** | `device_allocator<T>` wrapping `sycl::malloc_device`; custom `DeviceAllocator` concept | `thrust::device_allocator<T>` (CUDA `cudaMalloc`) | `device_allocator<T>` (`sycl::malloc_device`); supports alignment template parameter | USM: `sycl::usm_allocator<T, shared>` / Buffer: `__buffer_allocator<T>` |
| **Memory Model** | **Device memory** via `sycl::malloc_device`; host access triggers implicit transfers | **Device memory** via `cudaMalloc`; host access triggers implicit transfers | **Device memory** via `sycl::malloc_device`; implicit transfers | **Shared memory** via USM shared or SYCL buffer/accessor; runtime manages placement |

> | **Host Element Access** | `device_array`: explicit `host_read()`/`host_write()`; compat `device_vector`: `device_reference` proxy | Via `device_reference` proxy (explicit device-to-host copy) | Via `device_reference` proxy (`__SYCL_DEVICE_ONLY__` bifurcation) | Via `device_reference` proxy (runtime-managed migration) |

Reviewer: ```suggestion
| **Host Element Access** | `device_array`: explicit transfer to host; compat `device_vector`: `device_reference` proxy | Via `device_reference` proxy (explicit device-to-host copy) | Via `device_reference` proxy (`__SYCL_DEVICE_ONLY__` bifurcation) | Via `device_reference` proxy (runtime-managed migration) |
```

| **std::vector Interop** | Explicit constructor + `to_vector()` | Copy constructors from/to `std::vector` | Constructor from `std::vector` | Copy/move + implicit `operator std::vector()` |
| **Queue Association** | Stores context + device; queue provided per-operation or created on demand | Implicit (CUDA stream) | Allocator stores `device` + `context`; queue resolved at runtime via pointer introspection | Global default queue |
| **Uninitialized Construction** | `device_array`: uninitialized by default; compat `device_vector`: `no_init_t` tag | `default_init_t`, `no_init_t` tags | Not supported | Not supported |

## Proposal

The proposal consists of two complementary public types that share a
non-public base implementation, `internal::__device_storage_base`:

1. **[`device_array<T>`](device_array.md)** — the primary API.
   A clean, explicit, **fixed-size** container for device memory with no proxy

>    types. Explicit `host_read()`/`host_write()` for host access, uninitialized by

Reviewer: ```suggestion
   types. Explicit methods for host access or transfer, uninitialized by
```

   `operator[]` proxy semantics for drop-in migration from
   `thrust::device_vector`, along with a resizable, allocator-aware interface.

Both types **privately inherit** from
`internal::__device_storage_base<T, Alloc>`, which owns the shared machinery:
the device allocation and its lifetime, size, associated `sycl::context` /
`sycl::device`, the allocator instance, resizing, and the host-device transfer
helpers. Each derived type re-exposes (via `using` declarations) only the
subset of the base appropriate to its public contract — `device_array` omits
the allocator and resizing entirely, while `compat::device_vector` re-exposes
the full surface. This lets `device_array` present a simplified interface
without duplicating the implementation that `device_vector` reuses.

### Class Relationships

```mermaid
classDiagram
    direction LR

    namespace internal {
        class __device_storage_base~T, Alloc~ {
            owns allocation + size + context/device + allocator
            resize / host transfers
        }
    }

    namespace experimental {
        class device_array~T~ {
            fixed size, no allocator
            device access: span()

>             host access: host_read() / host_write()

Reviewer: ```suggestion
            host access: copy_to() / host_read() / host_write()
```

  alias `std::span` when `__cpp_lib_span >= 202002L` and fall back to
  `sycl::span` otherwise. Both are device copyable per SYCL 2020 §3.13.1, but
  preferring the standard type where it exists lets these spans compose with
  users' C++20 code and `std::ranges` without conversion. See
  [device_array](device_array.md#oneapidplspan).

- **No `push_back`, `insert`, `erase`.**
  Rarely used in practice (see [usage study](usage_pattern_study.md)),
  high implementation complexity for device memory.

- **Host-side operations block but do not synchronize with prior work.**
  The user is responsible for ensuring prior kernels have completed before
  host-side access. This can be achieved via an in-order queue or explicit
  event waits. No asynchronous overloads are proposed for either type; see
  [device_array](device_array.md#resolved-questions).

- **Header organization** - use individual headers <oneapi/dpl/device_array>,
  <oneapi/dpl/compat/device_vector>, we may add <oneapi/dpl/compat> in the
  future. This matches convention of the standard library and thrust.

- **Namespace** - device_array should be in `oneapi::dpl::experimental::device_array`.

>   The intention is to add `oneapi::dpl::compat` directly without `experimental`.

Reviewer: ```suggestion
  The intention for `device_vector` is to add it `oneapi::dpl::compat` directly without `experimental`.
```

  Alternatives:
  - Individual headers:
  `<oneapi/dpl/device_array>` and `<oneapi/dpl/device_vector>`.. `device_vector`
  would transitively include `device_array` since it depends on it.
  - We could have a `compat` header and a individual `device_array` header. If we intend to use `device_array` as convenient RAII device storage within oneDPL's sycl backend implementations, that may impact our decision here.

- **Compatibility namespace naming?** This proposal places the
  Thrust-compatible types in `oneapi::dpl::experimental::compat`. Several
  aspects are worth discussing:
  - Should `compat` be nested under `experimental`, or should it be
    `oneapi::dpl::compat` directly? Or alternatively `oneapi::dpl::ext::compat`.
  - Is `compat` the right name? Alternatives: `thrust_compat`, `migration`,
    `legacy`. `compat` is concise but doesn't indicate what it's compatible
    *with*. `thrust_compat` is more explicit but ties the namespace to a
    specific vendor's API.
  - Moreover, is this repository where we want the compatibility headers?  I think yes, otherwise they will be too cumbersome to use, but it worth raising.

> 

Reviewer 2: We need to look at what other projects have done regarding namespaces and be consistent I think.
* sycl-tla -- moved from `syclcompat` to `compat`
* SYCLomatic still maintains dpct and syclcompat namespaces, but I don't think this is important.

I would suggest we go directly to `oneapi::dpl::compat` rather than experimental for the device_vector.

Author: I've added the intention to go directly to `oneapi::dpl::compat`, rather than `oneapi::dpl::experimental::compat`.  Although, if there are more thoughts here we can discuss.  My intention is to tackle implementation of `device_array` in the short term, and `device_vector` in the longer term (not this upcoming milestone), so we have a bit of time to change our minds.

- **`device_reference` supports all compound assignment and increment/decrement operators**
  Following Thrust's convention, `device_reference<T>` will support all compound
  assignment operators (`+=`, `-=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `<<=`, `>>=`)
  and increment/decrement (`++`, `--`). This makes host-side element access behave
  as close to a real `T&` as possible and eases migration from Thrust codebases
  where users expect expressions like `d_vec[i] += 1` to work. Each such operation
  implies a synchronous round-trip (read-modify-write) to device memory.

- **No custom allocator template parameter; use `sycl::malloc_device` directly**
  The SYCL 2020 spec intentionally excludes `sycl::usm_allocator` for device
  memory because the C++ `Allocator` named requirement mandates host-accessible
  memory, which device USM is not. A device memory allocator cannot satisfy
  `Allocator`, so a custom allocator template parameter in the `std::vector`
  sense is not meaningful here. We use `sycl::malloc_device` / `sycl::free`
  directly, keeping the implementation simpler.

### API Skeleton

```cpp
namespace oneapi::dpl::experimental {

template <typename T>

> class device_vector {

Reviewer 3: What about allocator? At least I can use USM-shared and USM-device memory.
How we can specify that?

Reviewer: The RFC explicitly states that this choice is deliberately not provided. That can be discussed, of course. Also note that `sycl::usm_allocator` does not support `usm::alloc::device`.

Author: Yes, I believe the SYCL folks have deliberately not provided such an allocator because of some requirements on a C++ allocator which do not align with allocation on the device.  Generally, I think providing some generic allocator betrays the semantics here, as we always want the data to live on the device.  

Its worth thinking about if there are use cases which would benefit from a C++ allocator-like template argument which we defined to only provide USM device memory, but I don't have a convincing motivating case at this time.  

The best I can think of is some sort of debugging features like device memory bookkeeping, but it is a significantly more complex thing to add such a feature (and we could always extend it in the future).

public:
    // Standard container typedefs
    using value_type      = T;
    using size_type       = std::size_t;
    using difference_type = std::ptrdiff_t;
    using reference       = device_reference<T>;      // proxy
    using const_reference = device_reference<const T>; // proxy
    using pointer         = device_pointer<T>;
    using const_pointer   = device_pointer<const T>;
    using iterator        = /* device-aware random access iterator */;
    using const_iterator  = /* const version */;

    // Constructors / assignment / destructor -- mirrors std::vector
    // All constructors accept an optional sycl::queue argument currently,
    // but this is an open question.
    // (defaulted, e.g. to a global default queue).
    device_vector(sycl::queue q = /* default queue */);
    explicit device_vector(size_type count,
                           sycl::queue q = /* default queue */);
    device_vector(size_type count, const T& value,
                  sycl::queue q = /* default queue */);
    template <typename InputIt>
    device_vector(InputIt first, InputIt last,
                  sycl::queue q = /* default queue */);
    device_vector(std::initializer_list<T> init,

>                   sycl::queue q = /* default queue */);

Reviewer 3: At first look this default may cause a lot of problems: what is default queue?
For example if I create some `syql::queue` with CPU-selector this default will not work correctly together with it.
May be make sense to remove it at all?

Author: My expectation would be we would use the same as device policy's default queue `__internal::__get_default_queue()`.  

My understanding of what happens if you have provide a cpu-selector queue:
If the device does not support `aspect::usm_device_allocations`, `sycl::malloc_device` will throw `errc::feature_not_supported`, I think this is appropriate for a `device_vector`. 

For intel implementation of sycl, I believe that CPU devices *do* have that aspect (but I think it really depends upon the implementation of the Unified Runtime), and it will be equivalent to a malloc on the host.  Dereferencing this pointer on the "host" is UB according to the SYCL spec.  Currently, dereferencing a pointer to such an allocation on the "host" happens to work, but our implementation of `device_vector` will not rely upon that.  I think all this works as intended, with no special handling on our side.

    device_vector(const device_vector&); // also copies the queue
    device_vector(device_vector&&) noexcept;
    device_vector& operator=(const device_vector&);
    device_vector& operator=(device_vector&&) noexcept;
    ~device_vector();

    // Interop with std::vector
    explicit device_vector(const std::vector<T>&,
                           sycl::queue q = /* default queue */);
    explicit operator std::vector<T>() const;

    // Element access (proxy references, implies host-device transfer)
    reference       operator[](size_type pos);
    const_reference operator[](size_type pos) const;
    reference       front();
    const_reference front() const;
    reference       back();
    const_reference back() const;
    pointer         data() noexcept;
    const_pointer   data() const noexcept;

    // Iterators
    iterator begin() noexcept;
    iterator end()   noexcept;
    // + const/reverse variants

    // Queue access

>     sycl::queue get_queue() const noexcept;

Reviewer 3: May be make sense to return by reference?

Reviewer: `sycl::queue` is a common reference type, essentially a shared pointer.

Reviewer 3: OK

  4. **Full `std::vector`-like modifiers** (`push_back`, `insert`, `erase`)
     are rarely used in real workloads.
  5. **Raw pointer extraction** (`data().get()` / `raw_pointer_cast`) is
     ubiquitous for passing to kernels and library APIs.
  6. **`device_vector` as class member** is a major pattern in HPC
     (simulation state, sparse matrix storage, MPI buffers).

## Comparison of Existing Implementations

| Implementation | Source |
|---|---|
| **Thrust** (`thrust::device_vector`) | [NVIDIA/cccl - device_vector.h](https://github.com/NVIDIA/cccl/blob/main/thrust/thrust/device_vector.h) |
| **SYCLomatic** (`dpct::device_vector`) | [SYCLomatic - vector.h](https://github.com/oneapi-src/SYCLomatic/blob/SYCLomatic/clang/runtime/dpct-rt/include/dpct/dpl_extras/vector.h) |
| **Distributed Ranges** (`dr::sp::device_vector`) | [distributed-ranges - device_vector.hpp](https://github.com/oneapi-src/distributed-ranges/blob/main/include/dr/sp/device_vector.hpp) |
| **sycl-thrust** (`thrust::device_vector`) | [SparseBLAS/sycl-thrust - device_vector.h](https://github.com/SparseBLAS/sycl-thrust/blob/main/include/thrust/device_vector.h) |
### 1. How They Differ

> | Aspect | Thrust | SYCLomatic | Distributed Ranges | sycl-thrust | Proposed (oneDPL) |

Reviewer 4: Maybe put oneDPL column first. On my browser, the markdown rendered with a horizontal slider and I had to move it to see the oneDPL column. So perhaps order with oneDPL first and then (if an order of applies) from most to least interesting comparison.

Author: done.
